skbio/stats/composition.py

```diff
@@ -0,0 +1,384 @@
+
+    Returns
+    -------
+    array_like, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    """
+    mat = np.atleast_2d(mat)
+    if np.any(mat < 0):
+        raise ValueError("Cannot have negative proportions")
+    if mat.ndim > 2:
+        raise ValueError("Input matrix can only have two dimensions or less")
+    mat = mat / mat.sum(axis=1, keepdims=True)
+    return mat.squeeze()
+
+
+def multiplicative_replacement(mat, delta=None):
+    """Replace all zeros with small non-zero values.
+
+    It uses the multiplicative replacement strategy, replacing zeros
+    with small positive values and performing a closure operation so
+    that the compositions still add up to 1
+
+    Parameters
+    ----------
+    mat: array_like
+       a matrix of proportions where
+       rows = compositions and
+       columns = components
+
+    Returns
+    -------
+    numpy.ndarray, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skbio.stats.composition import multiplicative_replacement
+    >>> X = np.array([[.2,.4,.4, 0],[0,.5,.5,0]])
+    >>> multiplicative_replacement(X)
+    array([[ 0.1875,  0.375 ,  0.375 ,  0.0625],
+           [ 0.0625,  0.4375,  0.4375,  0.0625]])
+
+    """
+    mat = np.atleast_2d(mat)
+    if mat.ndim > 2:
+        raise ValueError("Input matrix can only have two dimensions or less")
+    if np.any(mat < 0):
+        raise ValueError("Cannot have negative proportions")
+    if not np.all(np.isclose(mat.sum(axis=1), 1)):
+        raise ValueError("Rows need to sum up to 1")
+
+    z_mat = (mat == 0)
+
+    num_samps, num_feats = mat.shape
+    tot = z_mat.sum(axis=1, keepdims=True)
+
+    if delta is None:
+        delta = (1. / num_feats)**2
+
+    zcnts = 1 - tot * delta
+    mat = np.where(z_mat, delta, zcnts * mat)
+    return mat.squeeze()
+
+
+def perturb(x, y):
+    r"""
+    Performs the perturbation operation.
+
+    This operation is defined as
+    :math:`x \oplus y = C[x_1 y_1, ..., x_D y_D]`
+
+    :math:`C[x]` is the closure operation defined as
+    :math:`C[x] = [\frac{x_1}{\sum x},...,\frac{x_D}{\sum x}]`
+    for some :math:`D` dimensional real vector :math:`x` and
+    :math:`D` is the number of components for every composition.
+
+    Parameters
+    ----------
+    x : array_like, float
+        a matrix of proportions where
+        rows = compositions and
+        columns = components
+    y : array_like, float
+        a matrix of proportions where
+        rows = compositions and
+        columns = components
+
+    Returns
+    -------
+    numpy.ndarray, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    Notes
+    -----
+    - All of the values in x and y must be non negative
```

Comment:
Or just remove the Notes section altogether, as I think every function requires the values to be nonnegative.